// sake/util.js
var Path = require('path');

module.exports = {
    /**
     * Convert a string into a JSON data structure, or just return the string
     * @param {String} src
     * @returns {Mixed}
     */
    jsonToValue: function (src) {
        var val;
        try {
            val = JSON.parse(src);
        }
        catch (e) {
            return src;
        }
        return val;
    },
    /**
     * Convert a string to a camel-cased version.  By default the input
     * string is split on whitespace, hyphens, or underscores, and each word
     * after the first is title-cased and then joined together with the
     * first.  i.e:
     * 
     *      "my-name-space" => "myNameSpace"
     *      "My-Class_name" => "MyClassName"
     * 
     * @param txt {string} the string to transform
     * @param splitDelim {RegExp|string} optional, where to split the string.
     *      Defaults to /[\s\-]/
     * @returns {string}
     */
    toCamelCase: function (txt, splitDelim) {
        var words = txt.split(splitDelim || /[\s\-]/);
        return words[0] + words.slice(1).map(function (w) {
            w = w.toLowerCase();
            return w[0].toUpperCase() + w.slice(1);
        }).join("");
    },
    
    STACK_PATH_REGEX: /\(([^:]+)(:\d+)+\)/,
    
    DEFAULT_STACK_LINE: 2,
    /**
     * Using an error stack, try to determine the calling filepath.
     * @param {Number} count optional additional stack lines to climb
     * @returns {string}
     */
    fileFromStackTrace: function (count) {
        var self = module.exports,
            error = new Error(),
            stack = error.stack.split("\n"),
            stackLine = self.DEFAULT_STACK_LINE + (count || 0),
            match = self.STACK_PATH_REGEX.exec(stack[stackLine]);
        
        return match && Path.resolve(match[1]) || process.cwd();
    },
    /**
     * As above, but returns the calling directory
     * @count {Number}
     * @returns {String}
     */
    directoryFromStackTrace: function (count) {
        // Don't use 'this' incase someone is just executing the function directly
        var self = module.exports;
        count = typeof count === 'number' && (count + 1) || 1;
        return Path.dirname(self.fileFromStackTrace(count));
    }
};
