// sake/file-utils.js

var Proteus = require("proteus"),
    FS      = require("fs"),
    Path    = require("path"),
    CP      = require("child_process"),
    sake    = require("sake"),
    Task    = require("sake/task"),

    DEFAULT_FILE_MODE       = "0644",
    DEFAULT_DIRECTORY_MODE  = "0777"
;
    
module.exports = {
    sh: function (cmd, success, failure) {
        sake.log("sh: " + cmd);
        Task.startAsync();
        CP.exec(cmd, function (error, stdout, stderr) {
            Task.clearAsync();
            if (!error && success) {
                success(error, stdout);
            }
            else if (error && success) {
                success(error, stderr);
            }
            else if (error) {
                sake.log(stderr);
            }
        });
    },

    mkdir: function (path, mode) {
        sake.log("mkdir " + path);
        FS.mkdirSync(path, mode);
    },
    
    mkdir_p: function (path, mode) {
        var cpath = path,
            missing = []
        ;
        
        sake.log("mkdir -p " + path);
        mode = mode || DEFAULT_DIRECTORY_MODE;
        while (!Path.existsSync(cpath) && !(cpath === "." || cpath === "/")) {
            missing.push(Path.basename(cpath));
            cpath = Path.dirname(cpath);
        }
        
        missing.reduceRight(function (t, c) {
            var p = Path.join(t, c);
            sake.mkdir(p, mode);
            return p;
        }, cpath);
    },

    rm: function () {
        util.slice(arguments, 0).forEach(function (path) {
            sake.log("rm " + path);
            FS.unlinkSync(path);
        });
    },

    rm_rf: function () {
        util.slice(arguments).forEach(function (path) {
            var stats;
            
            sake.log("rm -rf " + path);
            
            function filepath (f) {
                return Path.join(path, f);
            }

            if ((stats = FS.statSync(path)).isDirectory()) {
                sake.rm_rf.apply(
                    sake,
                    FS.readdirSync(path).map(filepath)
                );
                FS.rmdirSync(path);
            }
            else {
                sake.rm(path);
            }
        });
    },

    cp: function (from, to) {
        sake.log("cp " + from + " " + to);
        sake.write(to, sake.read(from));
    },

    mv: function (from, to) {
        sake.log("mv " + from + " " + to);
        FS.renameSync(from, to);
    },
    
    ln: function (from, to) {
        sake.log("ln " + from + " " + to);
        FS.linkSync(from, to);
    },

    ln_s: function (from, to) {
        sake.log("ln -s " + from + " " + to);
        FS.symlinkSync(from, to);
    },
    
    cat: function () {
        var args = util.slice(arguments);
        sake.log("cat " + args.join(" "));
        return args.map(function (path) {
            if (Array.isArray(path)) {
                return sake.cat.apply(sake, path);
            }
            return sake.read(path, "utf8");
        }).join("");
    },
    
    read: function (path, enc) {
        return FS.readFileSync.apply(FS, arguments);
    },
    
    write: function (/*path, data, enc, mode*/) {
        var args = util.slice(arguments, 0, 3),
            mode = arguments[3] || "w"
        ;
        if (mode === "a") {
            args[1] = args[1] + sake.read(args[0]);
        }
        FS.writeFileSync.apply(FS, args);
    }
};
