// sake/index.js

var Proteus         = require("proteus"),
    Path            = require("path"),
    nutil           = require("util"),
    taskDescription = "",
    sake,
    Task,
    FileTask,
    FileList,
    FileCreateTask
;
//---------------------------------------------------------------------------
// Privates
//---------------------------------------------------------------------------
function defineTask (TaskClass, name, prereqs, action) {
    var task = new TaskClass(name, prereqs, action);
    
    task.description = taskDescription;
    taskDescription = "";
    
    return task;
}
//---------------------------------------------------------------------------
// Public/Exports
//---------------------------------------------------------------------------
sake = {
    log: function () {
        var opts = this.options;
        
        if (!opts.quiet || opts.verbose) {
            Proteus.slice(arguments).forEach(function (arg) {
                arg = typeof arg === "string" ?
                    "[" + arg + "]" :
                    nutil.inspect(arg, false, 100);
                process.stderr.write(arg + "\n");
            });
        }
    },
    
    load: function (filepath) {
        require("sake/vm").run(filepath);
    },
    
    description: function (txt) {
        taskDescription += txt;
    },
    
    taskSync: function (name, deps, fn) {
        fn = fn ?
            function (t) {
                fn();
                t.done();
            } :
            null;
            
        return defineTask(Task, name, deps, fn);
    },
    
    directory: function () {
        
    }
};

module.exports = function (opts) {
    sake.options = opts || {};
    
    module.exports = sake;
    
    // Require these here to avoid recursive requires
    sake.Task           = Task              = require("sake/task");
    sake.FileTask       = FileTask          = require("sake/file-task");
    sake.FileCreateTask = FileCreateTask    = require("sake/file-create-task");
    sake.FileList       = FileList          = require("sake/file-list");
    // bind our task creation functions
    sake.task       = defineTask.bind(sake, Task);
    sake.file       = defineTask.bind(sake, FileTask);
    sake.fileCreate = defineTask.bind(sake, FileCreateTask);
    // some other functions that need late binding
    sake.desc = sake.description.bind(sake);
    sake.include = sake.load.bind(sake);
    
    return sake;
};
