// sake/task.js

var Proteus     = require("proteus"),
    defProp     = Object.defineProperty,
    defProps    = Object.defineProperties
;

module.exports = Task = Proteus.Class.derive({
    
    self: {
        lookup: function () {
            
        }
    },
    
    init: function (name, prereqs, action) {
        
        this.name = args.name;
        this.description = "";
        this.namespace = args.namespace;
        this.actions = [];
        this.enable();
        
        defProps(this, {
            __pending__: {
                value: true,
                writable: true
            },
            __pendingPrereqs__: {
                value: []
            },
            __prerequisites__: {
                value: []
            },
            prerequisites: {
                get: function () {
                    resolvePrerequisites.call(this);
                    return this.__prerequisites__;
                },
                enumerable: true
            }
        });
    }
});
