// sake/vm.js
var Proteus     = require("proteus"),
    VM          = require("vm"),
    Path        = require("path"),
    FS          = require("fs"),
    sake        = require("sake"),
    resolve     = require("resolve"),
    fileUtils   = require("sake/file-utils"),
    nodeRequire = require,
    sakeContext
;
//---------------------------------------------------------------------------
// Utility for resolving requires
//---------------------------------------------------------------------------
function sakeRequire (name) {
    try {
        return nodeRequire(name);
    }
    catch (e) {
        return nodeRequire(resolve.sync(name, {
            paths: sakeRequire.paths
        }));
    }
}

sakeRequire.paths = [__dirname, process.cwd()];
//---------------------------------------------------------------------------
// Create the Run Context
//---------------------------------------------------------------------------
sakeContext = VM.createContext(
    Proteus.merge(
        {
            console: console,
            require: sakeRequire,
            process: process
        },
        sake,
        fileUtils
    )
);
//---------------------------------------------------------------------------
// Exports
//---------------------------------------------------------------------------
module.exports = {
    
    run: function (filepath) {
        var code, ret;
        
        if (typeof filepath === "function") {
            code = "(" + filepath.toString() + "())";
            filepath = "anonymous function";
        }
        else {
            code = FS.readFileSync(filepath, "utf8");
            
            if (Path.extname(filepath).toLowerCase() === ".coffee") {
                code = require("coffee-script").compile(code);
                console.log(code);
            }
        }
        
        ret = VM.runInContext(code, sakeContext, filepath);
        return ret;
    }
    
};
